### PR TITLE
Unification of floating point literals definition

### LINKED ISSUE #2121: Float literal definition bug: suffix might appear in the literal twice
Here is the definition of floating point literal (covers decimal fp with radix separator and scientific representation):

https://github.com/ispc/ispc/blob/3231045399349438009d619724be2bea4baa5e65/src/lex.ll#L371

Note that `f` suffix appears twice in this definition, and this is a problem, as it effectively allows literals like `1.1fe+2f` or `1.1fp16e+2fp16`. This is clearly a bug and need to be fixed.

### PR DESCRIPTION
**Pull request by DeepakRajendrakumaran**
This commit does the following:
- `d` and `D` suffixes are accepted for scientific (`1.234e+3d`) and hexadecimal (`0x1.921fb54442d18p+1d`) notations.
- hexadecimal literals to accept `f16` suffix - the literal is treated as `float16` type.
- hexadecimal literals to accept both `p` and `P` (used to be just `p`): `0x1.1p+1` and `0x1.1P+1`. This is unification with C/C++, which accepts both.
- hexadecimal floating point literals and hex integer literals to accept both `0x` and `0X` prefixes (used to be only `0x`): `0X1.1p+1` and `0Xdeadbeaf`. This is unification with C/C++, which accepts both.
- fix #2121
- documentation improvement
- fix for `float16` constants handling - missing case in `ConstExpr::GetConstant()` / `ConstExpr::GetStorageConstant()`
- clean up for "Fortran double" lexical definition - unification with scientific notation (no user visible changes)
- report a specific error for `double` and `float16` literals without a radix separator, warn for `float` (`1f`). C/C++ do not allow that, so we match their behavior for `double`/`float16`, but not for `float`. The reason to allow it for `float` - it's there and we don't want to break existing users.

Before:
| | decimal | scientific | hexadecimal |
| --- | --- | --- | --- |
| float16 | **yes** | **yes** | --- |
| float | **yes** | **yes** | **yes** |
| double | **yes** | --- | --- |

After:
| | decimal | scientific | hexadecimal |
| --- | --- | --- | --- |
| float16 | **yes** | **yes** | **yes** |
| float | **yes** | **yes** | **yes** |
| double | **yes** | **yes**| **yes** |

### CODE REVIEW

File: docs/ispc.rst [hunk 1]
```diff
@@ -1587,40 +1592,59 @@ SI-inspired suffixes "k", "M", and "G" respectively:
...
+* The second option is scientific notation, where a base value is specified
+  as the first form of a floating-point constant but is then followed by an
+  "e" or "E", then a plus sign or a minus sign, and then an exponent.
+* Finally, floating-point constants may be specified as hexadecimal.
+  constants; this form can ensure a perfectly bit-accurate representation of
+  a particular floating-point number.  These are specified with an "0x"
+  prefix, followed by a zero or a one, a period, and then the remainder of
+  the mantissa in hexadecimal form, with digits from 0-9, a-f, or A-F.  The
+  start of the exponent is denoted by a "p", which is then followed by an
+  optional plus or minus sign and then digits from 0 to 9.
```
Comment (dbabokin) [inline docs/ispc.rst:1618]: What about this:
Hexadecimal floating-point constant - a hexadecimal constant (starting with "0x prefix) with or without radix separator, followed by "p" and a hexadecimal exponent value (without "0x" prefix). The exponent is never optional for hexadecimal floating-point literals.

**Review by dbabokin — COMMENTED** [review]
See my proposed changes. They are not just editorial. Differences:
- sign for the exponent is optional for both scientific and hexadecimal forms. I think this is already the way it is. This matches C++. Would be good to specifically test that.
- hexadecimal doesn't need to start with 1 or 0. This is the way it is in C++17. And it's not the way it is currently in ISPC. Should we match C++17 here?
- I've taken out the statement about "bit-accurate" representation for now. This needs additional consideration.

For bit-accurate statement it seems that ISPC notation allows only those numbers that can be represented in FP format without further normalization (1 or 0 before radix separator), while C++17 assumes normalization.
I'd like to understand if it's true for both ISPC and C++ formats that NaN / +-Inf / denormals / unnormal keep the bit pattern or not.

We need either have different (current ISPC) hexadecimal representation, stating that it keeps bit pattern and that it's different form C++ 17, or go c++17 way and specifically state what numbers keep bit patter and what are normalized.

**Comment by DeepakRajendrakumaran** [comment]
> See my proposed changes. They are not just editorial. Differences:
> 
> * sign for the exponent is optional for both scientific and hexadecimal forms. I think this is already the way it is. This matches C++. Would be good to specifically test that.

Yeah.. added test for this. It is supported.

File: docs/ispc.rst (hunk 1 above)
Comment (DeepakRajendrakumaran) [reply docs/ispc.rst]: I made this change. But as you pointed out - starting with 0 or 1 needs some investigation

File: docs/ispc.rst [hunk 2]
```diff
@@ -1587,40 +1592,59 @@ SI-inspired suffixes "k", "M", and "G" respectively:
...
 
-::
+All 3 floating-point constants can be specified in one of three ways.
+
+* First, they may be a sequence of zero or more digits from 0 to 9, followed
+  by a period, followed by zero or more digits from 0 to 9. (There must be at
+  least one digit before or after the period).
+* The second option is scientific notation, where a base value is specified
+  as the first form of a floating-point constant but is then followed by an
+  "e" or "E", then a plus sign or a minus sign, and then an exponent.
```
[the remark this answers is not in the document]
Comment (DeepakRajendrakumaran) [reply docs/ispc.rst]: Done

File: docs/ispc.rst [hunk 3]
```diff
@@ -1587,40 +1592,59 @@ SI-inspired suffixes "k", "M", and "G" respectively:
...
+* ``float16`` is a half-precision(16 bit format) floating point type.
+* ``float`` is a single-precision(32 bit format) floating point type.
+* ``double`` is a double-precision(64 bit format) floating point type.
 
-::
+All 3 floating-point constants can be specified in one of three ways.
+
+* First, they may be a sequence of zero or more digits from 0 to 9, followed
+  by a period, followed by zero or more digits from 0 to 9. (There must be at
+  least one digit before or after the period).
```
[the remark this answers is not in the document]
Comment (DeepakRajendrakumaran) [reply docs/ispc.rst]: Done ..but are we sure about this - 'If floating-point suffix is used, radix separator is optional.' - are you saying '8f' will be read as a float?

Looks like  another instance of difference between cpp and ispc:
https://godbolt.org/z/xK177TaGs - cpp
https://godbolt.org/z/73q4zbfEo - ispc

File: docs/ispc.rst [hunk 4]
```diff
@@ -1587,40 +1592,59 @@ SI-inspired suffixes "k", "M", and "G" respectively:
...
-prefix, followed by a zero or a one, a period, and then the remainder of
-the mantissa in hexadecimal form, with digits from 0-9, a-f, or A-F.  The
-start of the exponent is denoted by a "p", which is then followed by an
-optional plus or minus sign and then digits from 0 to 9.  For example:
+* ``float16`` is a half-precision(16 bit format) floating point type.
+* ``float`` is a single-precision(32 bit format) floating point type.
+* ``double`` is a double-precision(64 bit format) floating point type.
 
-::
+All 3 floating-point constants can be specified in one of three ways.
```
[the remark this answers is not in the document]
Comment (DeepakRajendrakumaran) [reply docs/ispc.rst]: Done

File: docs/ispc.rst [hunk 5]
```diff
@@ -1587,40 +1592,59 @@ SI-inspired suffixes "k", "M", and "G" respectively:
...
-Finally, floating-point constants may be specified as hexadecimal
-constants; this form can ensure a perfectly bit-accurate representation of
-a particular floating-point number.  These are specified with an "0x"
-prefix, followed by a zero or a one, a period, and then the remainder of
-the mantissa in hexadecimal form, with digits from 0-9, a-f, or A-F.  The
-start of the exponent is denoted by a "p", which is then followed by an
-optional plus or minus sign and then digits from 0 to 9.  For example:
+* ``float16`` is a half-precision(16 bit format) floating point type.
+* ``float`` is a single-precision(32 bit format) floating point type.
+* ``double`` is a double-precision(64 bit format) floating point type.
```
[the remark this answers is not in the document]
Comment (DeepakRajendrakumaran) [reply docs/ispc.rst]: Done

File: docs/ispc.rst [hunk 6]
```diff
@@ -1587,40 +1592,59 @@ SI-inspired suffixes "k", "M", and "G" respectively:
...
 
-  float16 one = 1e0f16;  // 1.0
-  float16 one  = 1.f16;  // 1.0
+  float two = 0x1p+1f;  // 2.0
+  float pi  = 0x1.921fb54442d18p+1f;  // 3.1415926535...
+  float neg = -0x1.ffep+11f;  // -4095.
+
+
+String Literals
+----------------
```
[the remark this answers is not in the document]
Comment (DeepakRajendrakumaran) [reply docs/ispc.rst]: Done

File: docs/ispc.rst [hunk 7]
```diff
@@ -100,6 +100,9 @@ Contents:
 
   + `Relationship To The C Programming Language`_
   + `Lexical Structure`_
+    * `Integer Literals`_
```
[the remark this answers is not in the document]
Comment (DeepakRajendrakumaran) [reply docs/ispc.rst:104]: Done

File: src/lex.ll [hunk 8]
```diff
@@ -512,20 +512,42 @@ L?\"(\\.|[^\\"])*\" { lStringConst(&yylval, &yylloc); return TOKEN_STRING_LITERA
...
+    std::string fp64C("D");
     std::string fp16S("f16");
     std::string fp16C("F16");
     if (val.size() >= fp16S.size() && ((val.compare(val.size() - fp16S.size(), fp16S.size(), fp16S) == 0)
            || (val.compare(val.size() - fp16C.size(), fp16C.size(), fp16C) == 0))) {
         yylval.stringVal = new std::string(val.substr(0, val.length() - 3));
         return TOKEN_FLOAT16_CONSTANT;
+    } else if (val.size() >= fp64S.size() && ((val.compare(val.size() - fp64S.size(), fp64S.size(), fp64S) == 0)
+           || (val.compare(val.size() - fp64C.size(), fp64C.size(), fp64C) == 0))) {
+        yylval.doubleVal = atof(yytext);
```
Comment (dbabokin) [inline src/lex.ll:525]: Any reason to do that through `atof` now, when we use APFloat to parse it on other code paths?

Also `(float)atof(yytext)` is bad practice as it parses as double and converts to float.

File: tests/lit-tests/float_literals.ispc [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    uniform double dval2 = 2.3D; // D suffix
+    uniform double dval3 = 2e+2d; // exponential form with signed exponential.
+    uniform double dval4 = 0x1.p3d; // normalized hex form with no sign.
+    uniform double dval5 = 2e2d; //exponential form with no sign for exponential.
+    uniform double dval6 = 0x1.p-3d; // normalized hex form with sign.
+    uniform double dval7 = 0x0.p+1d; // +0
+    uniform double dval8 = -0x0.p+1d; // -0
+    uniform double dval9 = +0x1.9p+8608d; // +infinity
+    uniform double dval10 = -0x1.9p+8608d; // -infinity
+    uniform double dval11 = 0x1.00001262d4ab6p-149d; // denormal
```
Comment (dbabokin) [inline tests/lit-tests/float_literals.ispc:83]: This still doesn't look like denormal. Is it?

**Comment by mko-io** [comment]
Since f16 has been native on new Intel Xe and AVX512 will have native fp16 support, why give fp16 a good short suffix like "h" for half

**Comment by dbabokin** [comment]
> Since f16 has been native on new Intel Xe and AVX512 will have native fp16 support, why give fp16 a good short suffix like "h" for half

I had this idea initially, but decided that `fp16` is a better choice because it's more explicit. Now when you are asking the question, I did some googling and see that `gcc` and `clang` for ARM target use suffix `f16` for `__fp16` and `_Float16` type. Somehow I missed that earlier.

I think it does make sense to switch to it to avoid confusion and divergence from `clang` and `gcc`. Using `h` suffix would just create divergence.

Any opinions?

File: docs/ispc.rst (hunk 1 above)
Comment (dbabokin) [reply docs/ispc.rst]: To be clear here - ISPC accepts hexadecimal literals with only 1 or 0 before radix separator, i.e. `0x1.fffp1` or `0x0.fffp1`, which means the fractional part is exactly mantissa, without any modifications. This form is subset of C++17 hexadecimal floating point literals (see [here](https://en.cppreference.com/w/cpp/language/floating_literal) for details), which accept any hexadecimal value and do normalization.

I think it makes sense to accept C++17 literals, especially if/when we switch to APFloat for parsing literals. But his should be another change set. Let's keep current restriction for now.

File: tests/lit-tests/float_literals.ispc (hunk 9 above)
Comment (dbabokin) [reply tests/lit-tests/float_literals.ispc]: I've fixed that.

**Comment by dbabokin** [comment]
The changes are ready for final review.

**Review by aneshlya — APPROVED** [review]
LGTM!

**Pull request merged**